src/base_maths.jl

```diff
@@ -76,23 +76,30 @@ function Base.:*(B1::BlockDiagonal, B2::BlockDiagonal)
     end
 end
 
-function _check_matmul_dims(A::AbstractMatrix, B::AbstractMatrix)
+function _check_matmul_dims(A::AbstractMatrix, B::AbstractVecOrMat)
     # match error message from LinearAlgebra
     size(A, 2) == size(B, 1) || throw(DimensionMismatch(
         "A has dimensions $(size(A)) but B has dimensions $(size(B))"
     ))
 end
 
-function Base.:*(B::BlockDiagonal{T}, M::AbstractMatrix) where T
-    _check_matmul_dims(B, M)
+_mulblocksizes(bblocks, ::AbstractVector) = size.(bblocks, 1)
+_mulblocksizes(bblocks, M::AbstractMatrix) =
+    zip(size.(bblocks, 1), Base.Iterators.repeated(size(M, 2), length(bblocks)))
```

Comment:
following [Blue style](https://github.com/invenia/BlueStyle#method-definitions) let's use a long-form method definition when it doesn't fit on one line:

```julia
function _mulblocksizes(bblocks, M::AbstractMatrix)
    return zip(size.(bblocks, 1), Base.Iterators.repeated(size(M, 2), length(bblocks)))
end
```